hastory-server/src/Hastory/Server/OptParse.hs

```diff
@@ -0,0 +1,155 @@
+import Data.Maybe
+import qualified Data.Text as T
+import Data.Yaml
+import qualified Env
+import GHC.Generics (Generic)
+import Options.Applicative as OptParse
+import qualified Options.Applicative.Help as OptParse (string)
+import Path
+import Path.IO
+import Text.Read
+import YamlParse.Applicative as YamlParse
+
+data Instructions = Instructions Dispatch Settings deriving (Show, Eq, Generic)
+
+newtype Dispatch = DispatchServe ServeSettings deriving (Show, Eq, Generic)
+
+data Settings = Settings deriving (Show, Eq, Generic)
+
+data ServeSettings = ServeSettings {serveSettingsPort :: Int, serveSettingsLogFile :: Path Abs File, serveSettingsKeyFile :: Path Abs File} deriving (Show, Eq, Generic)
+
+data Arguments = Arguments Command Flags deriving (Show, Eq, Generic)
+
+newtype Command = CommandServe ServeArgs deriving (Show, Eq, Generic)
+
+data ServeArgs = ServeArgs {serveArgsPort :: Maybe Int, serveArgsLogFile :: Maybe FilePath, serveArgsKeyFile :: Maybe FilePath} deriving (Show, Eq, Generic)
+
+newtype Flags = Flags {flagsConfigFile :: Maybe FilePath} deriving (Show, Eq, Generic)
+
+data Environment = Environment {envPort :: Maybe Int, envLogFile :: Maybe FilePath, envKeyFile :: Maybe FilePath, envConfigFile :: Maybe FilePath} deriving (Show, Eq, Generic)
+
+data Configuration = Configuration {configPort :: Maybe Int, configLogFile :: Maybe FilePath, configKeyFile :: Maybe FilePath} deriving (Show, Eq, Generic)
+
+instance FromJSON Configuration where
+  parseJSON = viaYamlSchema
+
+instance YamlSchema Configuration where
+  yamlSchema = objectParser "Configuration" $ Configuration <$> optionalField "port" "Port to start server on." <*> optionalField "log" "File to save logs to." <*> optionalField "key" "File to read / write JWK key from / to."
+
+getInstructions :: IO Instructions
+getInstructions = do
+  args@(Arguments _ flags) <- getArguments
+  env <- getEnvironment
+  config <- getConfiguration flags env
+  combineToInstructions args env config
+
+getArguments :: IO Arguments
+getArguments = customExecParser prefs_ argParser
+
+getEnvironment :: IO Environment
+getEnvironment = Env.parse (Env.header "Environment") environmentParser
+
+combineToInstructions :: Arguments -> Environment -> Maybe Configuration -> IO Instructions
+combineToInstructions (Arguments cmd _flags) Environment {..} mConf = Instructions <$> dispatch <*> settings
+  where
+    dispatch =
+      case cmd of
+        CommandServe ServeArgs {..} -> do
+          let serveSettingsPort = fromMaybe 8080 (serveArgsPort <|> envPort <|> mc configPort)
+          serveSettingsLogFile <-
+            maybe defaultLogFile resolveFile' (serveArgsLogFile <|> envLogFile <|> mc configLogFile)
+          serveSettingsKeyFile <-
+            maybe defaultKeyFile resolveFile' (serveArgsKeyFile <|> envKeyFile <|> mc configKeyFile)
+          pure $ DispatchServe (ServeSettings {..})
+    settings = pure Settings
+    mc f = mConf >>= f
+
+defaultKeyFile :: IO (Path Abs File)
```

Comment:
Nit: We usually call functions that query something from inside IO `getX` and the results `x`. So here that would be `getDefaultKeyFile` and the result would be `defaultKeyFile`.

Reply:
Done!